### ISSUE TITLE
Question about thread safety

### ISSUE THREAD

**Issue by jmattheis** [rogchap/v8go#290]
The documentation of `v8go.NewIsolate()` says Isolates aren't thread safe

https://github.com/rogchap/v8go/blob/d8d94c25bd2b2bb5b785d6e184a4562901d165da/isolate.go#L47-L49

Promise#then says `Context#PerformMicrotaskCheckpoint` must be executed until the promise is resolved.

https://github.com/rogchap/v8go/blob/d8d94c25bd2b2bb5b785d6e184a4562901d165da/promise.go#L93-L95

When one goroutine executes `Context#PerformMicrotaskCheckpoint` and another one uses the isolate for v8.Null() or v8.NewValue() to create the resolve promise value, does this create thread safety problems, because the isolate is used in two different goroutines? How can I find out, if something can be used in different goroutines like v8.NewValue?

Here simple program to reproduce, it is similar to the example in the PR that added support for promises https://github.com/rogchap/v8go/pull/76#issue-806262194

<details>
  <summary>Simple Program</summary>
  
```go
func main() {
	iso := v8.NewIsolate()
	sleep := v8.NewFunctionTemplate(iso, func(info *v8.FunctionCallbackInfo) *v8.Value {
		ctx := info.Context()
		p, err := v8.NewPromiseResolver(ctx)
		if err != nil {
			panic(err)
		}
		go func() {
			time.Sleep(1 * time.Second)
			p.Resolve(v8.Null(iso))
		}()
		return p.Value
	})
	global := v8.NewObjectTemplate(iso)
	global.Set("sleep", sleep)
	ctx := v8.NewContext(iso, global)
	_, err := ctx.RunScript(`const test = async () => { await sleep(); };`, "sleep.js") // executes a script on the global context
	if err != nil {
		panic(err)
	}
	val, err := ctx.RunScript("test()", "exec.js")
	if err != nil {
		panic(err)
	}
	p, _ := val.AsPromise()

	p.Then(func(info *v8.FunctionCallbackInfo) *v8.Value {
		fmt.Println("promise done")
		return v8.Null(iso)
	})
	fmt.Println("waiting")
	for p.State() == v8.Pending {
		ctx.PerformMicrotaskCheckpoint()
	}

	fmt.Println("done done")
}
````

</details>

**Comment by dylanahsmith** [comment]
It seems like the documentation has already answered your question, so your example isn't supported.

Instead of resolving the promise from another goroutine, send the result over a channel to resolver it from the isolate's goroutine.

The one exception is [Isolate.TerminateExecution](https://pkg.go.dev/rogchap.com/v8go#Isolate.TerminateExecution) which is intended to be used from another goroutine, since it is needed in order to interrupt JS execution (e.g. to enforce a deadline).  It looks like that is missing from the documentation at the moment.

**Comment by jmattheis** [comment]
thanks.

**Comment by jmattheis** [comment]
@dylanahsmith Is the example at https://pkg.go.dev/rogchap.com/v8go#example-FunctionTemplate-Fetch faulty, because it uses the iso (for creating the string(body) outside of the goroutine that has created the isolate?

**Comment by dylanahsmith** [comment]
Good point.

For context, v8go does internally use v8::Locker on the isolate, which provides some type safety. However, the primary reason for its use is to handle the fact that goroutines switch which thread they are run on and taking that lock will get v8 to update a thread-local stack limit so that V8 can handle the thread switch.

I think that example of using (*PromiseResolver).Resolve probably would work given the current use of v8::Locker, but I'm not sure if that would change in the future.  However, v8go also has wrapper Go structs around V8 data which aren't type safe.  That means relying on that could result in breaking changes when upgrading v8go, such as from state being moved between C++ structs and Go structs (e.g. as part of addressing https://github.com/rogchap/v8go/issues/105).

It seems like that example should be modified to not rely the method being type safety to reflect what is actually supported.  I'll reopen this since this does reflect an issue with the documentation being inconsistent.

There is another issue with more discussion about thread-safety: https://github.com/rogchap/v8go/issues/129

**Comment by rutsky** [comment]
> It seems like the documentation has already answered your question, so your example isn't supported.
> 
> Instead of resolving the promise from another goroutine, send the result over a channel to resolver it from the isolate's goroutine.

So you **can't** call [`PromiseResolver.Resolve()`](https://pkg.go.dev/rogchap.com/v8go#PromiseResolver.Resolve) from a separate goroutine that could potentially ran in parallel with other javascript code for the same Isolate in V8?

And examples like

* in the first comment in this thread,
* https://github.com/rogchap/v8go/blob/cee5f84bb54338bdef150b3e43daace4669c5de2/function_template_fetch_test.go#L33-L37
* `fetch()` implementation in v8go-polyfills by @kuoruan : https://github.com/kuoruan/v8go-polyfills/blob/c74c5b408ebdd76b65a26c693cf6cd612b2dace7/fetch/fetcher.go#L96-L101 

are buggy and likely cause eventual crashes?
